valhalla/baldr/time_info.h

```diff
@@ -0,0 +1,226 @@
+      return {false};
+    }
+
+    // Find the first edge whose end node has a valid timezone index and keep it
+    int timezone_index = 0;
+    for (const auto& pe : location.path_edges()) {
+      const baldr::GraphTile* tile = nullptr;
+      const auto* edge = reader.directededge(baldr::GraphId(pe.graph_id()), tile);
+      timezone_index = reader.GetTimezone(edge->endnode(), tile);
+      if (timezone_index != 0)
+        break;
+    }
+
+    // Set the origin timezone to be the timezone at the end node use this for timezone changes
+    if (timezone_index == 0) {
+      // Don't use the provided one if its not valid
+      if (!dt::get_tz_db().from_index(default_timezone_index)) {
+        default_timezone_index = baldr::DateTime::get_tz_db().to_index("Etc/UTC");
+      }
+      LOG_WARN("No timezone for location using default");
+      timezone_index = default_timezone_index;
+    }
+    const auto* tz = dt::get_tz_db().from_index(timezone_index);
+
+    // Get the current local time and set it if its a current time route
+    // NOTE: we set to minute resolution to match the input format
+    const auto now_date =
+        date::make_zoned(tz, sc::time_point_cast<sc::seconds>(
+                                 sc::time_point_cast<sc::minutes>(sc::system_clock::now())));
+    if (location.date_time() == "current") {
+      std::ostringstream iso_dt;
+      iso_dt << date::format("%FT%R", now_date);
+      location.set_date_time(iso_dt.str());
+    }
+
+    // Convert the requested time into seconds from epoch
+    date::local_seconds parsed_date;
+    try {
+      parsed_date = dt::get_formatted_date(location.date_time(), true);
+    } catch (...) {
+      LOG_ERROR("Could not parse provided date_time: " + location.date_time());
+      return {false};
+    }
+    const auto then_date = date::make_zoned(tz, parsed_date);
+    uint64_t local_time = date::to_utc_time(then_date.get_sys_time()).time_since_epoch().count();
+
+    // What second of the week is this (for historical traffic lookup)
+    auto second_of_week = dt::second_of_week(local_time, tz);
+
+    // When is this route with respect to now this will let us appropriately use current flow traffic
+    int64_t seconds_from_now = (then_date.get_local_time() - now_date.get_local_time()).count();
+
+    // Make a valid object
+    return {true,
+            static_cast<uint64_t>(timezone_index),
+            local_time,
+            second_of_week,
+            static_cast<uint64_t>(std::abs(seconds_from_now)),
+            seconds_from_now < 0,
+            tz_cache};
+  }
+
+  /**
+   * Offset all the initial time info to reflect the progress along the route to this point
+   * @param seconds_offset  the number of seconds to offset the TimeInfo by
+   * @param next_tz_index   the timezone index at the new location
+   * @return a new TimeInfo object reflecting the offset
+   */
+  inline TimeInfo forward(float seconds_offset, int next_tz_index) const {
```

Comment:
this method and its sister method below are in charge of updating the timeinfo object to reflect the passage of time as expansion proceeds. so after you traverse an edge you provide the amount of time that passed during that traversal to this function so it can update its members to reflect that time change.